File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/history/PackageTrackScreen.kt
package lk.mageride.passenger.history

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Call
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.map.MageRideMap
import lk.mageride.passenger.map.MapCamera
import lk.mageride.passenger.map.MapPin
import lk.mageride.passenger.map.MapVehicle
import lk.mageride.passenger.map.VehicleLayers
import lk.mageride.passenger.ride.CallChoice
import lk.mageride.passenger.ride.CallChooserSheet
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme

/**
 * SCR-PA-020 and SCR-PA-021 — the parcel, from whichever end is holding the phone.
 *
 * The wireframes: a map with the driver's live marker, a **four-step bar** (`Pickup pending ·
 * Picked up · In transit · Delivered`), the party's own OTP, and the driver card with a Call.
 *
 * **Which of the two this is, is a fact about the ride and not about the link.**
 * `mageride://package/{rideId}` is the same URI for both parties — the sender gets it on
 * `package_delivered`, the recipient on `package_picked_up` — so the view model reads the ride and
 * decides. See [PackageParty].
 *
 * **A recipient never signed in to get here** (P-09, AL-45): they came from an FCM deep link, or —
 * with no app — from an SMS onto the SCR-WT web page, which is not this app's problem.
 */
@Composable
@Suppress("LongMethod") // The wireframe's layout tree: map, step bar, OTP, driver card.
internal fun PackageTrackScreen(
    onBack: () -> Unit,
    onFreeCall: () -> Unit,
    choice: CallChoice,
    model: PackageTrackViewModel,
) {
    val state by model.state.collectAsStateWithLifecycle()
    var callOpen by remember { mutableStateOf(false) }

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = stringResource(
                    if (state.party == PackageParty.SENDER) {
                        R.string.package_track_sending
                    } else {
                        R.string.package_track_incoming
                    },
                ),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }

        Box(modifier = Modifier.weight(1f)) {
            MageRideMap(
                modifier = Modifier.fillMaxSize(),
                vehicles = state.driverPosition?.let {
                    listOf(MapVehicle(vehicleId = DRIVER_MARKER, lat = it.lat, lng = it.lng))
                }.orEmpty(),
                pins = listOfNotNull(
                    state.ride?.pickup?.let { MapPin(VehicleLayers.PIN_PICKUP, it.lat, it.lng) },
                    state.ride?.dropoff?.let { MapPin(VehicleLayers.PIN_DROPOFF, it.lat, it.lng) },
                ),
                camera = state.driverPosition?.let { MapCamera(it.lat, it.lng) } ?: MapCamera.Default,
            )
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(
                    MaterialTheme.colorScheme.background,
                    RoundedCornerShape(topStart = MageRideTheme.radius.card, topEnd = MageRideTheme.radius.card),
                )
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            StatusBar(step = state.step)

            OtpRow(state = state)

            state.ride?.driver?.let { driver ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(
                            text = driver.name,
                            style = MaterialTheme.typography.bodyLarge,
                            color = MaterialTheme.colorScheme.onSurface,
                        )
                        Text(
                            text = listOfNotNull(
                                driver.registrationNumber,
                                driver.rating?.let { stringResource(R.string.ride_rating, it) },
                            ).joinToString(SEPARATOR),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    OutlinedButton(onClick = { callOpen = true }) {
                        Icon(
                            imageVector = Icons.Filled.Call,
                            contentDescription = null,
                            modifier = Modifier.size(ControlTokens.ChipIcon),
                        )
                        Text(
                            text = stringResource(R.string.ride_call),
                            modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                        )
                    }
                }
            }

            state.error?.let { InlineError(message = stringResource(it)) }
        }
    }

    if (callOpen) {
        CallChooserSheet(
            driverName = state.ride?.driver?.name,
            // AL-48 — the real number, from `Accepted` onward. On a package `counterpartyPhone` is
            // the FAR end, which for a sender is the recipient; the driver is who this screen calls,
            // so it takes the ride's driver block and the number the contract carries.
            driverPhone = state.ride?.counterpartyPhone,
            choice = choice,
            onFreeCall = {
                callOpen = false
                onFreeCall()
            },
            onDialled = { },
            onDismiss = { callOpen = false },
        )
    }
}

/** The wireframe's four-step bar (US-20.7). */
@Composable
private fun StatusBar(step: Int) {
    val labels: List<Int> = listOf(
        R.string.package_step_pending,
        R.string.package_step_picked_up,
        R.string.package_step_in_transit,
        R.string.package_step_delivered,
    )

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
    ) {
        labels.forEachIndexed { index, label ->
            Column(modifier = Modifier.weight(1f), horizontalAlignment = Alignment.CenterHorizontally) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(ControlTokens.SheetHandleHeight)
                        .background(
                            if (index <= step) {
                                MaterialTheme.colorScheme.primary
                            } else {
                                MaterialTheme.colorScheme.surfaceVariant
                            },
                            RoundedCornerShape(MageRideTheme.radius.sm),
                        ),
                )
                Text(
                    text = stringResource(label),
                    modifier = Modifier.padding(top = MageRideTheme.spacing.xxs),
                    style = MaterialTheme.typography.labelSmall,
                    color = if (index <= step) {
                        MaterialTheme.colorScheme.onSurface
                    } else {
                        MaterialTheme.colorScheme.onSurfaceVariant
                    },
                    textAlign = TextAlign.Center,
                )
            }
        }
    }
}

/**
 * The party's own handover code.
 *
 * **Or an honest absence.** Neither OTP can be read back from the platform — the pickup code is
 * returned once at booking and the delivery code arrives only on a push — so a screen opened
 * without one has nothing to show, and says so rather than drawing four empty boxes that look like
 * a loading state. See [PackageOtps].
 */
@Composable
private fun OtpRow(state: PackageTrackState) {
    if (state.delivered) return

    val label = if (state.party == PackageParty.SENDER) {
        R.string.package_pickup_otp_label
    } else {
        R.string.package_delivery_otp_label
    }

    Column {
        Text(
            text = stringResource(label),
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = state.otp ?: stringResource(R.string.package_otp_unavailable),
            style = if (state.otp != null) {
                MaterialTheme.typography.headlineMedium
            } else {
                MaterialTheme.typography.bodyMedium
            },
            color = if (state.otp != null) {
                MaterialTheme.colorScheme.onSurface
            } else {
                MaterialTheme.colorScheme.onSurfaceVariant
            },
        )
    }
}

private const val SEPARATOR = " · "
private const val DRIVER_MARKER = "driver"

File: apps/passenger-ios/PassengerApp/History/PackageTrackScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-020 and SCR-PI-021 — the parcel, from whichever end is holding the phone.
///
/// Both cells: a short map with the driver's live marker, the **four-step bar** (`Pickup pending ·
/// Picked · Transit · Delivered`), this party's own handover code in a `card fill`, and the driver
/// row underneath.
///
/// **Which of the two this is, is a fact about the ride and not about the link.**
/// `mageride://package/{rideId}` is the same URI for both parties — the recipient gets it on
/// `package_picked_up`, the sender on `package_delivered` — so the model reads the ride and decides.
/// See ``PackageParty``.
///
/// **The two cells differ in three places and each difference is drawn.** The sender's header is
/// *"Package · sending"* and the recipient's is *"Incoming package"*; the sender's code is the
/// **pickup** OTP and the recipient's is the **delivery** OTP; and the sender's driver row carries a
/// **📞 Call** where the recipient's carries the **ETA**. The last is the wireframe's own decision and
/// is followed here — `apps/passenger-android` draws a Call on both, which is a difference recorded
/// in the C099 handoff rather than reproduced.
@MainActor
struct PackageTrackScreen: View {

    @StateObject private var model: PackageTrackModel

    /// SCR-PI-015a's *"Free call"* — the VoIP screen (C102).
    let onFreeCall: () -> Void

    private let choice: CallChoice
    private let contact: RideContact

    @State private var isCallOpen = false

    init(
        rideId: String,
        history: HistoryRepository,
        live: PassengerLiveMap,
        otps: PackageOtps,
        signedInUserId: String?,
        choice: CallChoice,
        contact: RideContact,
        onFreeCall: @escaping () -> Void
    ) {
        _model = StateObject(
            wrappedValue: PackageTrackModel(
                rideId: rideId,
                history: history,
                live: live,
                otps: otps,
                signedInUserId: signedInUserId
            )
        )
        self.choice = choice
        self.contact = contact
        self.onFreeCall = onFreeCall
    }

    var body: some View {
        VStack(spacing: 0) {
            map

            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                StepperBar(titleKeys: PackageTrackState.stepKeys, current: model.state.step)

                if let headline = statusHeadline {
                    Text(headline)
                        .mageFont(.bodyEmphasis)
                        .foregroundStyle(MageRideColor.onSurface)
                }

                // Nothing to read out once the parcel has changed hands — the card would be a code
                // for a handover that already happened.
                if !model.state.isDelivered {
                    HandoverCodeCard(titleKey: codeTitleKey, code: model.state.otp)
                }

                driverRow

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                Spacer(minLength: 0)
            }
            .padding(MageRideSpacing.md)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: titleKey))
        .navigationBarTitleDisplayMode(.inline)
        .task { model.start() }
        .onDisappear { model.stop() }
        .sheet(isPresented: $isCallOpen) {
            CallChooserSheet(
                driverName: model.state.ride?.driver?.name,
                driverRating: model.state.ride?.driver?.rating?.doubleValue,
                // AL-48's clear number. **On a package ride this field is the far *party*, not the
                // driver** — see ``PackageTrackModel`` and the gap the C099 handoff records; it is
                // what the contract gives a participant and what the Android twin dials.
                driverPhone: model.state.driverPhone,
                choice: choice,
                contact: contact,
                onFreeCall: {
                    isCallOpen = false
                    onFreeCall()
                },
                onDismiss: { isCallOpen = false }
            )
        }
    }

    // MARK: -

    /// The cell's `.map` strip — the driver's marker, and the two ends of the delivery.
    ///
    /// The camera follows the **driver** while there is one, because that is the thing that moves;
    /// before the first `DriverPosition` it opens on the drop-off, which is where the parcel is going.
    private var map: some View {
        MageRideMap(
            vehicles: driverMarker,
            pins: pins,
            camera: camera
        )
        .allowsHitTesting(false)
        .frame(height: MageRideControl.packageMapHeight)
    }

    /// SCR-PI-021's `In transit · ~12 min`.
    ///
    /// Drawn only where the cell draws it — on the recipient's screen, and only once there is an ETA
    /// to state. The sender's cell puts the step name in the bar and nothing above the card.
    private var statusHeadline: String? {
        guard model.state.party == .recipient, let minutes = model.state.etaMinutes else { return nil }
        let step = PackageTrackState.stepKeys[min(model.state.step, PackageTrackState.deliveredStep)]
        return step.localised + MageRideSymbols.separator + "ride_arriving_minutes".localisedFormat(Int(minutes))
    }

    /// The driver, and what each party may do about them.
    @ViewBuilder
    private var driverRow: some View {
        if let driver = model.state.ride?.driver {
            HStack(spacing: MageRideSpacing.xs) {
                DriverIdentityRow(
                    name: driver.name,
                    // The sender's cell draws `K. Fernando · Tuk ★4.7`; the recipient's draws
                    // `K. Fernando · Tuk · ABC-1234` over `ETA 12 min`.
                    rating: model.state.party == .sender ? driver.rating?.doubleValue : nil,
                    vehicle: VehicleToken.forType(driver.vehicleType)?.nameKey.localised,
                    plate: model.state.party == .recipient ? driver.registrationNumber : nil,
                    etaMinutes: model.state.party == .recipient ? model.state.etaMinutes : nil
                )

                // **The Call is the sender's only** — the recipient's cell draws none. See the
                // type's note.
                if model.state.party == .sender {
                    TextLink(key: "ride_call") { isCallOpen = true }
                }
            }
        }
    }

    private var titleKey: String {
        model.state.party == .sender ? "package_track_sending" : "package_track_incoming"
    }

    private var codeTitleKey: String {
        model.state.party == .sender ? "package_pickup_otp_label" : "package_delivery_otp_label"
    }

    /// The driver's marker, as one frame the map tweens (MAP-04).
    ///
    /// The vehicle's *type* is passed so the marker is drawn in §0.2's legend colour rather than in
    /// the fallback grey — the same rule MAP-03 makes about every other marker in this app.
    private var driverMarker: [MapVehicle] {
        guard let position = model.state.driverPosition else { return [] }
        return [
            MapVehicle(
                vehicleId: PackageTrackScreen.driverMarkerId,
                lat: position.lat,
                lng: position.lng,
                type: model.state.ride?.driver?.vehicleType?.wire
            ),
        ]
    }

    private var pins: [MapPin] {
        guard let ride = model.state.ride else { return [] }
        return [
            MapPin(kind: VehicleLayers.pinPickup, lat: ride.pickup.lat, lng: ride.pickup.lng),
            MapPin(kind: VehicleLayers.pinDropoff, lat: ride.dropoff.lat, lng: ride.dropoff.lng),
        ]
    }

    private var camera: MapCamera {
        if let position = model.state.driverPosition {
            return MapCamera(lat: position.lat, lng: position.lng)
        }
        guard let dropoff = model.state.ride?.dropoff else { return .colombo }
        return MapCamera(lat: dropoff.lat, lng: dropoff.lng)
    }

    /// The one marker this screen ever draws. `MageRideMap` keys its source on the id, and a ride's
    /// driver has no `vehicleId` on `RideDriver` at all.
    private static let driverMarkerId = "package-driver"
}
